### ISSUE TITLE
Dependency management

### ISSUE THREAD

**Issue by adrianchifor** [deepmind/kapitan#91]
We could of course use plain git submodules, but what if we only need only one file or folder from an external git repo, http url or even helm chart/templates ?

I would like to introduce the concept of modular dependencies in kapitan, and it would look something like this:

```
parameters:
  kapitan:
    compile:
    ...

    dependencies:
    - type: git
      output_path: lib
      source: git@github.com:grafana/grafonnet-lib.git
      subdir: grafonnet
      version: 0.1
    - type: http
      output_path: lib
      source: https://raw.githubusercontent.com/bitnami-labs/kube-libsonnet/master/kube.libsonnet
      version: 0.1
```

In this scenario, `kapitan compile` will pull `github.com/grafana/grafonnet-lib/grafonnet/*` and `github.com/bitnami-labs/kube-libsonnet/master/kube.libsonnet` into `lib` at the root. Dependencies will be cached and only re-pulled when the **version** changes, keeps things simple and we don't need to implement an extra kapitan subcommand.

Users have full control of what the **source** can be (jsonnet, libsonnet, jinja2 templates, k8s manifests etc.) and where it's placed inside the project (**output_path**), can be per target, class or global.

In terms of modularity, each **type** can be defined as a separate python module with the same name, as in `kapitan/kapitan/dependencies/git.py` will control the functionality of `type: git`; in the example above, **subdir** is a git module specific variable, we can also add branch, release etc.

@ademariag @gburiola @ramaro What do you think?

**Comment by sbarzowski** [comment]
FYI there's an effort to create a standard dependency management for jsonnet, called jsonnet-bundler. I think the authors have thought it out pretty well. I see you want to do deal with much more than just jsonnet, so I'm not sure how relevant it is.

repo: https://github.com/jsonnet-bundler/jsonnet-bundler
doc: https://docs.google.com/document/d/1czRScSvvOiAJaIjwf3CogOULgQxhY9MkiBKOQI1yR14/edit?usp=sharing

**Comment by adrianchifor** [comment]
Yes we've seen that, we can easily create a dependency module specifically for jsonnet (`type: jsonnet`) that uses the standard. The idea is still very fresh, once it's a bit more mature we will take a look :)

**Comment by adrianchifor** [comment]
@sbarzowski Forgot to mention, we merged #133 a while ago, which allows the usage of jsonnet-bundler to pull external dependencies and reference them inside kapitan.

**Comment by caioaao** [comment]
My 2C on the subject: I don't think kapitan be responsible for managing these kinds of dependencies as it will needlessly grow in complexity since it will be managing dependencies of other tools, like jsonnet or Python, and we know the latter is already a mess with a plethora of tools and different dependency management tools.

A more generic and simpler approach IMO would be having pre-conditions for compilation. This way the user is responsible for its build environment, but kapitan is able to check if the needed tools are found.

tldr: if kapitan is already free from managing Python dependencies, why should it manage jsonnet ones (or any other, for that matter)?

**Comment by nikkomega** [comment]
I believe (correct me if I'm wrong, anyone) this issue isn't about build/runtime type dependencies like you're describing, but having kapitan configuration that's dependent on other repositories or sources of kapitan (or relevant jsonnet/jinja snippets, etc) config, i.e., having a some-app config repo that references an external standard-myorg-app-base as a parent, and/or pulls in a common envoy-sidecar one as a mixin.

**Comment by harsh-98** [comment]
@adrianchifor.  As mentioned by caioaao we don't want a complex dependencies manager. Only a checker which find whether we are using specific version  or not. If not, tell user to use that version.
```python
    dependencies:
    - type: jsonnet
      module: kube
      version: 0.12
```
Also,  just to clarify, if a user has installed kapitan with jsonnet version 0.11 and is using python 2.7. So, if user try run kapitan on main.jsonnet with version specific as 0.12, kapitan should raise warning. Right?

**Comment by adrianchifor** [comment]
@harsh-98 Jsonnet versions are already managed in `requirements.txt` and you can pin kapitan to a version by specifying `version: <kapitan version>` in `.kapitan`. We need a way to pull external dependencies like jsonnet/kadet libs and helm charts from git repos or http endpoints, bring them in and manage the versioning of them to know when to check/re-pull them.

**Comment by caioaao** [comment]
And that's exactly what I'm arguing against. Kapitan should manage kapitan dependencies, not jsonnet/Kadet/helm/whatever is next. Each of those are external tools and, as such, should have external dependency managers. Kapitan will become impossible to maintain otherwise

**Comment by ramaro** [comment]
@caioaao we're not talking about supporting dependency management for external tools.
This is for kapitan depedencies. For example, this would simplify importing a project from another repository into your current template repository as @nikkomega mentioned above.

**Comment by adrianchifor** [comment]
We had https://github.com/deepmind/kapitan/issues/197 for allow custom scripts during compilation but now that https://github.com/deepmind/kapitan/pull/267 is merged, people can write and import custom jinja filters. Also, running scripts on compile would not be able to keep track of dependencies, like when to re-pull. Yes https://github.com/deepmind/kapitan/issues/143 is a subtask of this.

**Comment by yoshi-1224** [comment]
@adrianchifor @uberspot 
Just looking at #290, do you think it makes sense to support copy of local files via this feature?

**Comment by yoshi-1224** [comment]
@uberspot yeah, just as another input type `file` (or some other appropriate name) on top of `http[s]` and `git`.

**Comment by uberspot** [comment]
File:// is defined in rfcs https://en.m.wikipedia.org/wiki/File_URI_scheme so it's a good contender for that. sounds good to me otherwise. :)

**Comment by prakashrj** [comment]
Is there any reason, why this issue is still open after #304 is merged?

**Comment by ramaro** [comment]
@prakashrj the only reason is that we forgot to close! Thanks :)
